### ISSUE TITLE
Segfault and negative years when string-formating datetimes with large years

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#3589]
Trying to print (or display in a notebook) a datetime with a very large year segfaults, or gives a negative year:

```python
import scipp as sc

print(sc.datetime('2147483647-12-04T15:30:07'))  # ok
print(sc.datetime('2147483648-12-04T15:30:07'))  # formats as negative year!
print(sc.datetime('2200000000-12-04T15:30:07'))  # segfault
```

**Comment by elindgren** [comment]
Is this related to max integer in scipp? I played around a bit with the `datetime64` object in numpy for comparison, and it behaves similarly with negative datetimes for large inputs (however no segfaults).

```python
np.datetime64(np.iinfo('int').max-1970, 'Y')  # numpy.datetime64('9223372036854775807'), ok
np.datetime64(np.iinfo('int').max-1969, 'Y')  # (numpy.datetime64('NaT'), ok
np.datetime64(np.iinfo('int').max-1968, 'Y')  #  (numpy.datetime64('-9223372036854775807'), negative year
```

Interestingly, creating a datetime with more than max int number of days will raise an error, as opposed to a negative time
```python
~ np.datetime64(np.iinfo('int').max+1, 'D'), np.iinfo('int').max
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[110], line 1
----> 1 np.datetime64(np.iinfo('int').max+1, 'D'), np.iinfo('int').max

OverflowError: int too big to convert

```

**Comment by jl-wynen** [comment]
I don't think this is directly related to max ints. E.g., this still segfaults:
```python
str(sc.datetime(np.iinfo('int').max-1000, unit='s'))
```
But creating the datetime object works fine. (Also without the `-1000`.) So I suspect this is a problem with the formatter.

**Comment by jl-wynen** [comment]
I tried to write a test for this but got strange results. I wrote this test:
```cpp
TEST_F(ISODateTest, large_time) {
  const core::time_point time{67767976231169407l};
  EXPECT_EQ(to_iso_date(time, units::s), "2147483647-12-04T15:30:07");
}

```
where the large integer represents 2147483647-12-04T15:30:07 in seconds since epoch.
I have two code bases for Scipp, one for use in CLion and one for PyCharm. The setup should be identical. But the above test fails in CLion because the date is rendered as 2176-03-12 04:19:23. But it passes in PyCharm. So far, I could not find a difference between the two that would explain this. The CMake arguments are the same.

Drilling down into the implementation of `to_iso_date`, I think the 'wrong' output of 2176-03-12 04:19:23 makes sense. We convert the `scipp::core::time_point` to a `std::chrono::system_clock`. The latter stores its time as *nano*seconds since epoch. And that conversion overflows, giving us the wrong date. I don't know why this doesn't happen in my PyCharm build. Or apparently with a released Scipp package as shown in the OP.

I don't know if this relates to the segfault or negative date. But if it does, we need to find a way to render a datetime without going through `std::chrono::system_clock`.
